Create Haddocks from entity documentation comments

blujupiter32 opened the pull request:
This PR adds a flag to `MkPersistSettings` which enables Haddock generation from entity documentation comments. The default is `False` to avoid any unexpected performance impacts from upgrading `persistent`.

Fixes #1462.

---

Before submitting your PR, check that you've:

- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock
  - Changes to the documentation assume that this feature will be made available in `persistent` version `2.14.6.0`.
- [x] Ran `stylish-haskell` on any changed files.
  - `stylish-haskell` fails on the `#if`-`#else`-`#endif` construct in the import list in `TH.hs` as all three preprocessor directives are ignored, resulting in two identifier lists for `Language.Haskell.TH.Lib`.
- [x] Adhered to the code style (see the `.editorconfig` file for details)

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Bumped the version number if there isn't an `(unreleased)` on the Changelog
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

<!---Thanks so much for contributing! :)

_If these checkboxes don't apply to your PR, you can delete them_-->

parsonsmatt commented:
This is fantastic, thanks!

Can you clarify the limitatiion around documenting fields?

Can you post the problematic test? That smells like a GHC bug to me, and might be worth reporting upstream.

For a test, I'd be happy to see a rendered Haddock that contains the docs - a screenshout would be fine.

blujupiter32 commented:
On closer inspection, the limitation only applies to `withDecDoc`, so I've added Haddock support for fields using `putDoc`. As for the test, I've added a spec module in https://github.com/blujupiter32/persistent/pull/5 which fails with the error. Building the Haddock documentation for `persistent` from this branch should show an additional `Test.HaddockExample` module with entries generated from the doc comments.

parsonsmatt commented on persistent/test/Database/Persist/TH/EntityHaddockSpec.hs:
```diff
@@ -0,0 +1,36 @@
+#if MIN_VERSION_template_haskell(2,18,0)
+import Database.Persist.TH.CommentSpec (CommentModel (..))
+import Language.Haskell.TH (DocLoc (DeclDoc), getDoc)
+import Language.Haskell.TH.Lift (lift)
```
```suggestion
import Language.Haskell.TH.Syntax (lift)
```

Should work to allow removing the `th-lift` dependency

blujupiter32 replied:
Fixed in 3fbb38362faba6465d4894bc64ae0ec907fc1ee8.

blujupiter32 commented on persistent/persistent.cabal:
```diff
@@ -145,6 +145,7 @@ test-suite test
       , silently
       , template-haskell         >= 2.4
       , text
+      , th-lift
```
Fixed in 3fbb38362faba6465d4894bc64ae0ec907fc1ee8.

blujupiter32 commented:
Regarding version bumping, would this change require a minor version bump (to `2.15.0.0`) as `mpsEntityHaddocks` has been added to the public API?

Also, would it be worth addressing #1296 and #1297 (and merging #1505) before releasing this change so that documentation comments are syntactically equivalent to the format supported by Haddock? Alternatively, these issues could be fixed and released in subsequent patch versions if frequent releases aren't a problem.

parsonsmatt commented:
This would be a minor version bump, which put be `2.14.6.0`, since PVP has two major versions.

I think I'd be fine releasing this separately from changing the way the comment syntax works.

blujupiter32 commented:
e2420bc78130dc6614b517860fe0a09abb6f6ecb conflicts with 0bc89423a6c7452a6345a9a1e33e63c92d5d9d19, so would you prefer rebasing onto `master` or merging `master` into the feature branch? PRs seem to be squash-merged so I don't think it would make much difference either way.

parsonsmatt commented:
You can do either; I generally find that merges make viewing diffs slightly easier on GitHub since I can see "changes since last review," but you're right that it'll get squash/merged into `master` anyway.
